NameSwitch

johnhaddon opened the pull request:
This adds the NameSwitch node as per #2853. The ArrayPlug changes contain a small ABI change, so can't be backported to 0.54 or 0.53 as-is. We can use a nasty `IECore::ClassData` approach if we do want to backport them, but I didn't want that in master.

johnhaddon commented:
Note that I didn't add extra spacing between the default plug and the others. I think NoduleLayout can support it as-is, but since the spacing is already pretty wide I wasn't sure it'd read all that well. Should I drop the spacing between the other plugs, or introduce a little vertical dividing line? Or not worry?

themissingcow commented:
> Note that I didn't add extra spacing between the default plug and the others. I think NoduleLayout can support it as-is, but since the spacing is already pretty wide I wasn't sure it'd read all that well. Should I drop the spacing between the other plugs, or introduce a little vertical dividing line? Or not worry?

It feels like it works OK as is, but then I know how its meant to work...  might be interesting to see it with 4px extra between the default and the others?

johnhaddon commented:
> It feels like it works OK as is, but then I know how its meant to work... might be interesting to see it with 4px extra between the default and the others?

I looked into this and it's a bit trickier than I thought. You can register custom gadgets and position them anywhere in the nodule layout, but because we have a high spacing between nodules, we also get a high spacing between our little custom 4px spacer gadget and the nodules - hence a _very_ high spacing between the default input and the next.

themissingcow commented:
> > It feels like it works OK as is, but then I know how its meant to work... might be interesting to see it with 4px extra between the default and the others?
> 
> I looked into this and it's a bit trickier than I thought. You can register custom gadgets and position them anywhere in the nodule layout, but because we have a high spacing between nodules, we also get a high spacing between our little custom 4px spacer gadget and the nodules - hence a _very_ high spacing between the default input and the next.

Seems like its worth running with it as is and see how people get on with it?

danieldresser-ie commented:
Took a look through.  Conceptually, making a bunch of changes to Switch to allow NameSwitch to use it as a base class makes me kind of uncomfortable - I generally have a lot of sympathy for the view that implementation inheritance is evil, and this seems like a textbook example where the two end up very tightly bound.

But looking at the actual code in this case, it actually ends up looking pretty tidy, and we're probably fine with Switch and NameSwitch being extremely tightly coupled ( anyone outside of Gaffer trying to do something like this would be more troublesome ).

So, LGTM.

I do have a question though:  do you expect promoting these fancy array plugs to work?  I'm thinking about the use case where we want to make an IERenderPassSceneSwitch, which is just a NameSwitch with ScenePlugs where the selector is automatically hardcoded to read from a context  variable named something like "ieRenderPass".

The best way to do this would probably be a box with an expression driving the selector and the inputs and outputs promoted.  Hopefully that works, the alternative I guess would be inheriting from NameSwitch.

johnhaddon commented:
> Conceptually, making a bunch of changes to Switch to allow NameSwitch to use it as a base class makes me kind of uncomfortable - I generally have a lot of sympathy for the view that implementation inheritance is evil, and this seems like a textbook example where the two end up very tightly bound.

Agreed. I was somewhat uncomfortable too, but I wanted to implement things in a backportable way, and this was the only way I could see, short of a completely standalone node and even more special cases in GraphLayout, ShaderPlug etc.

> But looking at the actual code in this case, it actually ends up looking pretty tidy, and we're probably fine with Switch and NameSwitch being extremely tightly coupled ( anyone outside of Gaffer trying to do something like this would be more troublesome ).

Yep, most of the changes to Switch are legit bug fixes for things that could affect Switch itself. What makes me uncomfortable are the non-virtual "overriding" of `setup()` and the driving of the pretend-its-not-public `indexPlug()`. One reasonable approach to tidying all that up would be to add a `SwitchBase` class with pure virtual `setup()` and `activeIndex()` methods and then derive Switch and NameSwitch from it separately. Happy to look into that if you think its a better approach and don't mind targeting this at 0.55.

> I do have a question though: do you expect promoting these fancy array plugs to work? 

Promoting plugs with fancy resizing-from-within-signals behaviours is something I'm really wary of, but the new "don't resize when inputs change" behaviour is an improvement in that respect. I'm hoping to eventually make that the norm. In a quick test it does seem that the new UI components I've added don't like being promoted, but that should be a reasonably easy fix. I may not get to that today though...

murrays-ie commented:
Thanks John! A little user-feedback:

The drag to reorder behaviour feels a bit unbalanced. To reorder a plug below its immediate neighbour you are only required to drag it down a few pixels to the top of the neighbouring row. To reorder a plug above its immediate neighbour requires dragging it above the full height of the neighbouring row. How about we adjust this so to reorder in either direction you need to drag past the half-way point of the neighbouring row so you're dragging about the same distance in either direction?

Enabling/disabling plugs does not propagate dirtyness, you need to change the frame or disable/enable the node to correctly see the result of disabling a plug.

A small complaint, but with the way the UI is arranged in the Node Editor the Default plug label is vertically aligned with the column of enable checkboxes below it, which leads me to look at the UI and think "what's the purpose of these 'Default' checkboxes on every plug?" Could we instead align the "Default" label with the plug name text widgets so each row is more visually consistent?

themissingcow commented:
> A small complaint, but with the way the UI is arranged in the Node Editor the Default plug label is vertically aligned with the column of enable checkboxes below it, which leads me to look at the UI and think "what's the purpose of these 'Default' checkboxes on every plug?"

That confused me for a moment too. Another option could be to move the checkbox to the beginning of the row? As it disables everything other than itself, having it first may feel more intuitive?

themissingcow commented:
> > A small complaint, but with the way the UI is arranged in the Node Editor the Default plug label is vertically aligned with the column of enable checkboxes below it, which leads me to look at the UI and think "what's the purpose of these 'Default' checkboxes on every plug?"
> 
> That confused me for a moment too. Another option could be to move the checkbox to the beginning of the row? As it disables everything other than itself, having it first may feel more intuitive?

I just noticed though, this would then break the pattern used by `CustomAtttributes`, etc.. Maybe using the toggle switch instead of a checkbox would help?

johnhaddon commented:
@murrays-ie :

> The drag to reorder behaviour feels a bit unbalanced.

Fixed in 7a1966e.

> Enabling/disabling plugs does not propagate dirtyness, you need to change the frame or disable/enable the node to correctly see the result of disabling a plug.

Fixed 67a9be31cdbfdf1f293a0521eec139899157fcad

> the way the UI is arranged in the Node Editor the Default plug label is vertically aligned with the column of enable checkboxes below it, which leads me to look at the UI and think "what's the purpose of these 'Default' checkboxes on every plug?" Could we instead align the "Default" label with the plug name text widgets so each row is more visually consistent?

Fixed in dad6783.

@themissingcow :

> Maybe using the toggle switch instead of a checkbox would help?

Done in e54f4d3.

johnhaddon commented:
I've rebased to fix the merge conflicts, and pushed 1618622 as a best-effort at supporting promotion without major surgery to the metadata system. As noted in the comments, it's not entirely satisfactory, but it's the best I can come up with right now.

danieldresser-ie commented:
1618622 is a tad ugly, but it seems like a reasonable way to get current support in for a feature that we do need.  I'd say LGTM.
